plugins/wu_palmer.py

```diff
@@ -267,10 +277,154 @@ def get_requirements(self) -> str:
-    # get parameters
+def load_taxonomy_as_node_paths(taxonomy: Dict) -> Dict[str, Tuple[str, ...]]:
+    nodes: Dict[str, Tuple[str, ...]] = {}
+    edges: Dict[str, str] = {}
+    root_node: Optional[str] = ""
+
+    if taxonomy["type"] != "tree":
+        TASK_LOGGER.warn("taxonomy is not a tree")
+        return None
+
+    for relation in taxonomy["relations"]:
+        if relation["target"] != relation["source"]:
+            edges[relation["target"]] = relation["source"]
+
+    for node in taxonomy["entities"]:
+        node_id: str
+
+        if isinstance(node, str):
+            node_id = node
+        elif isinstance(node, dict):
+            node_id = node["ID"]
+        elif isinstance(node, tuple):
+            node_id = node[0]
+        else:
+            TASK_LOGGER.warn("entity has an unsupported type")
+            continue
+
+        ancestors = [node_id]
+        visited = {node_id}
+        current_node_id = node_id
+
+        for _ in range(len(edges)):
+            current_node_id = edges.get(current_node_id, None)
+
+            if current_node_id is None:
+                break
+
+            if current_node_id in visited:
+                TASK_LOGGER.warn(
+                    f"cycle detected, node {current_node_id} already visited"
+                )
+                return None
+
+            ancestors.append(current_node_id)
+            visited.add(current_node_id)
+
+        if root_node is not None:
+            if root_node == ancestors[-1]:
+                root_node = None
+
+            if not root_node:
+                root_node = ancestors[-1]
+
+        nodes[node_id] = tuple(ancestors)[::-1]
+
+    return nodes
+
+
+class WuPalmerCache:
+    def __init__(self, taxonomies: Dict[str, Dict], root_has_meaning_in_taxonomy: bool):
+        self._taxonomies = taxonomies
+        self._node_path_cache: Dict[str, Dict[str, Tuple[str, ...]]] = {}
+        self._root_has_meaning_in_taxonomy = root_has_meaning_in_taxonomy
+
+    @lru_cache
+    def calculate_similarity(self, tax_name: str, node_a: str, node_b: str) -> float:
+        if tax_name not in self._node_path_cache:
+            self._node_path_cache[tax_name] = load_taxonomy_as_node_paths(
+                self._taxonomies[tax_name]
+            )
 
+        node_paths = self._node_path_cache[tax_name]
+
+        if node_a not in node_paths:
+            raise RuntimeError(f"node {node_a} not in node paths of taxonomy {tax_name}")
+
+        if node_b not in node_paths:
+            raise RuntimeError(f"node {node_b} not in node paths of taxonomy {tax_name}")
+
+        ancestors_a = node_paths[node_a]
+        ancestors_b = node_paths[node_b]
+
+        common_ancestor_depth = 0 if self._root_has_meaning_in_taxonomy else -1
+
+        for a, b in zip(ancestors_a, ancestors_b):
+            if a != b:
+                break
+
+            common_ancestor_depth += 1
+
+        if self._root_has_meaning_in_taxonomy:
```

Comment:
use an if at the beginning to set a variable `root_node_depth` to 0 or 1 then always subtract that depth from the denominator (x2) and the common ancestor depth (before the loop). This way it would be more clear on what is going on without more comments.